Expose child routes on currentRoute object; add property for a CSS class on inactive routes

tmcdos opened the pull request:
It is not easy to get access to child routes when you need to render nested menus with `<router-link>`
Publishing **children** on the **$route.currentRoute** solves a real problem. The **parent** property is required so that the route can see its siblings. The **title** property is optional - it is useful instead of **name** when you want to name a route having a default child. I personally use **name** in `beforeRouteEnter/beforeRouteUpdate` to update the **document.title** in browser - and I need it on all routes (even with default child), so I had to add **title** property.

This closes issue #1149

There are non-trivial situations where playing with CSS selectors becomes too inconvenient. Consider the following markup:
```vue
    <div>
      <div class="tabs_account">
        <div class="tab_spacer">&nbsp;</div>
        <template v-for="page in $router.currentRoute.parent.children" v-if="page.can_show">
          <router-link tag="div" active-class="tab_active" inactive-class="tab" v-bind:to="page.path">{{ page.title }}</router-link>
          <div class="tab_spacer">&nbsp;</div>
        </template>
        <div class="tab_fill tab_spacer">&nbsp;</div>
      </div>
      <router-view></router-view>
    </div>
```
You can achieve the same effect like this
```vue
<router-link :class="{'router-link-active': $route.fullPath ==='/' || $route.fullPath === '/a', 'router-link-inactive': $route.fullPath !=='/' && $route.fullPath !== '/a'}" to="/a">
Text
</router-link>
```
but it looks hacky (you have to keep the PATH strings in sync between HTML and JavaScript code) while the proposed solution feels more elegant.

tmcdos committed:
feat(router-link): add 2 properties for a CSS class on inactive routes (linkInactiveClass, linkExactInactiveClass)

There are non-trivial situations where playing with CSS selectors becomes too inconvenient. Consider the following markup:
    <div>
      <div class="tabs_account">
        <div class="tab_spacer">&nbsp;</div>
        <template v-for="page in $router.currentRoute.parent.children" v-if="page.can_show">
          <router-link tag="div" active-class="tab_active" inactive-class="tab" v-bind:to="page.path">{{ page.title }}</router-link>
          <div class="tab_spacer">&nbsp;</div>
        </template>
        <div class="tab_fill tab_spacer">&nbsp;</div>
      </div>
      <router-view></router-view>
    </div>
You can achieve the same effect like this
<router-link :class="{'router-link-active': $route.fullPath ==='/' || $route.fullPath === '/a'}" to="/a"></router-link>
but it looks hacky (you have to keep the PATH in HTML in sync with the JavaScript code) while my proposed solution feels more elegant.

tmcdos committed:
feat(core): publish 3 new properties on $route.currentRoute (children, parent, title)

It is not easy to get access to child routes when you need to render nested menus with <router-link>
Publishing "children" on the $route.currentRoute solves a real problem. The "parent" property is required so that the route can see its siblings. The "title" property is optional - it is useful instead of "name" when you want to name a route having a default child. I personally use "name" in beforeRouteEnter/beforeRouteUpdate to update the "document.title" in browser - and I need it on all routes (even with default child), so I had to add "title" property.

This closes issue #1149

posva commented:
Hey thanks for the PR 🙂 
If you want to add the feature, please, only add **that** feature, so don't add the `title` feature which is a personal thing. Also, please, don't add the `// TMCDOS` at the end of every line you added
Remember to add a test case for children

tmcdos committed:
feat(core): expose parent and child routes on currentRoute, add property for CSS class on inactive router links

The $router.currentRoute object now exposes "children[]" and "parent" properties.
The <router-link> component has now properties "inactive-class" and "exact-inactive-class" for specifying CSS class on inactive routes.

Solves issue #1149

tmcdos commented:
Okay, I followed your recommendations:

- removed the `// TMCDOS` marks
- removed the custom **title** property
- added a unit test for **parent** and **children[]** properties
- added a Selenium test for **inactive-class** and **exact-inactive-class**
- fixed some settings to allow headless browser testing with ChromeDriver

Hopefully the PR is now acceptable.

posva commented on package.json:
```diff
@@ -28,7 +28,7 @@
     "test": "npm run lint && flow check && npm run test:unit && npm run test:e2e && npm run test:types",
     "test:unit": "jasmine JASMINE_CONFIG_PATH=test/unit/jasmine.json",
-    "test:e2e": "node test/e2e/runner.js",
+    "test:e2e": "xvfb-run node test/e2e/runner.js && rm -rf ./xvfb-run.* && rm -rf ./.com.google.* && rm -rf ./.org.chromium.*",
```
Please let it as it was

posva commented on test/e2e/nightwatch.config.js:
```diff
@@ -33,7 +33,10 @@ module.exports = {
         'browserName': 'chrome',
         'javascriptEnabled': true,
-        'acceptSslCerts': true
+        'acceptSslCerts': true,
```
Can you remove this as well please

posva commented on test/unit/specs/children.spec.js:
```diff
@@ -0,0 +1,78 @@
+Vue.use(VueRouter)
+
+describe('currentRoute', () => {
+  describe('children[]', () => {
```
I don't think a describe is necessary here. a test with _should have children_ and another with _should have parent.children_ is enough

posva reviewed (changes requested):
I thought you were going to remove the other extra feature as well 😅 . Can you remove the inactive class feature from this PR, please?
Also, could you add the typings to the `flow/declarations.js` file, please?

Frondor commented:
I believe this is overkill, child routes should never be exposed in `currentRoute` object. A new method like `this.$route.getChildren()` should be better in my opinion.

tmcdos commented:
Child routes are exposed in every **route** object - just like the **parent** property. Why do you think a method is better than a property ?

Frondor commented:
Actually, no. They aren't.
The parent route is just one referenced object, and if you are exposing children for those route objects, you're wasting a lot of memory for something that most people don't need. That's why **in my opinion** it should be called/built on demand instead.

tmcdos commented:
Forgive me my profanity - perhaps using the original JavaScript objects as defined when instantiating the router instead of partially cloning them and exposing these partial clones would solve the problem with memory waste ?

kanatkubash commented:
Are not js objects just referenced not copied, it shouldnt be wasting memory.

posva commented:
Closing in favour of #1991
